File: src/gofire/command/command.go
// Package command provides commands for querying both, server and client.
package command

import (
	"encoding/json"
	"gofire/message"
	"gofire/user"
)

type CommandType int

const (
	REGISTER CommandType = iota //Register a user
	//Text-Messaging Commands
	BMESSAGE //Broadcast-Message
	UMESSAGE //Unicast-Message
	MMESSAGE //Multicast-Message
	//System Commands
	BLOGIN  //Broadcast-Login
	BLOGOUT //Broadcast-Logout
)

//Every Command has a type and a value for sending arbitrary data.
type Command struct {
	Type  CommandType
	Value []byte
}

//This Function wraps a message in a command.
//This is a shortcut for Wrap.
//If something went wrong, nil and a error are returned.
func PrepareMessage(tp CommandType, usr *user.User, msg []byte) (*Command, error) {
	mMessage, err := json.Marshal(message.Message{User: usr, Msg: msg})
	if err != nil {
		return nil, err
	}

	return &Command{tp, mMessage}, nil
}
//This Function wraps everything into a Command.
//It returns either nil and a error or a command pointer and nil
func Wrap(tp CommandType, v interface{}) (*Command, error){
	mStruct, err := json.Marshal(v)
	if err != nil {
		return nil, err
	}

	return &Command{tp, mStruct}, nil
}

File: src/gofire/main/template/js/command.js
const REGISTER = 0 //Register a user
//Text-Messaging Commands
const BMESSAGE = 1 //Broadcast-Message
const UMESSAGE = 2 //Unicast-Message
const MMESSAGE = 3 //Multicast-Message
	
const BLOGIN = 4 //Broadcast-Login
const BLOGOUT = 5 //Broadcast-Logout

function Command(type, value){
	this.Type = type;
	this.Value = value;
}

function Encode(obj){
	if(obj.Type && obj.Value){
		return new Command(obj.Type, obj.Value);
	}
	return 'undefined';
}

Command.prototype.Send = function(websocket){
	this.value = window.btoa(this.value);
	websocket.send(JSON.stringify(this));
}